packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -123,9 +123,84 @@ class LineBoundary extends TextBoundary {
   TextRange getTextBoundaryAt(int position) => _textLayout.getLineAtOffset(TextPosition(offset: max(position, 0)));
 }
 
+/// A text boundary that uses paragraphs as logical boundaries.
+///
+/// A paragraph is defined as the range between line terminators. If no
+/// line terminators exist then the paragraph boundary is the entire document.
+class ParagraphBoundary extends TextBoundary {
+  /// Creates a [ParagraphBoundary] with the text.
+  const ParagraphBoundary(this._text);
+
+  final String _text;
+
+  /// Returns the [int] representing the start position of the paragraph that
+  /// bounds the given `position`. The returned [int] is the position of the code unit
+  /// that follows the line terminator that encloses the desired paragraph.
+  @override
+  int? getLeadingTextBoundaryAt(int position) {
+    if (position < 0 || position > _text.length) {
```

Comment:
Wouldn't we be unable to iterate through the text if the given position is not contained within the text? Or in that case do we set the position to the end of the text and start iterating from there?

Reply:
Eot is also considered a paragraph break I think. The convention is to return the largest index regardless of the bounds, but I don't know if there're significant benefits over returning null when the index is OOB.

Reply:
I think if someone where to use `ParagraphBoundary` a stand-alone class with a text of 'hello world'. If they use `100` for the position given to `getLeadingBoundaryAt` and `getTrailingBoundaryAt`, then they would expect null bounds since `position` `100` does not exist within the text.

Reply:
```dart
  /// Returns the offset of the closest text boundary before or at the given
  /// `position`, or null if no boundaries can be found.
  ///
  /// The return value, if not null, is usually less than or equal to `position`.
  int? getLeadingTextBoundaryAt(int position);
```
This is the definition so by that definition it should return the smallest offset, "returning null if OOB" currently is not a promise the interface makes, I think it's OK to make that change but to keep it consistent all subclasses would need to change their behavior. But there doesn't seem to be a huge benefit that you gain from switching from one to another?

Reply:
Oh okay that makes sense. Sorry I was misunderstanding before, I am fine with keeping the definition with how it is. I have updated the implementation to reflect this.